ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/ProtocolSchedule.java

```diff
@@ -12,17 +12,43 @@
  *
  * SPDX-License-Identifier: Apache-2.0
  */
+
 package org.hyperledger.besu.ethereum.mainnet;
 
+import org.hyperledger.besu.ethereum.core.BlockHeader;
+import org.hyperledger.besu.ethereum.core.BlockHeaderBuilder;
 import org.hyperledger.besu.ethereum.core.ProcessableBlockHeader;
 
-public interface ProtocolSchedule
-    extends HeaderBasedProtocolSchedule, PrivacySupportingProtocolSchedule {
+import java.math.BigInteger;
+import java.util.Optional;
+import java.util.function.Predicate;
+
+public interface ProtocolSchedule extends PrivacySupportingProtocolSchedule {
 
-  ProtocolSpec getByBlockNumber(long number);
+  ProtocolSpec getByBlockHeader(final ProcessableBlockHeader blockHeader);
 
-  @Override
-  default ProtocolSpec getByBlockHeader(final ProcessableBlockHeader blockHeader) {
-    return getByBlockNumber(blockHeader.getNumber());
+  default ProtocolSpec getForNextBlockHeader(
+      final BlockHeader parentBlockHeader, final long timestampForNextBlock) {
+    final BlockHeader nextBlockHeader =
+        BlockHeaderBuilder.fromHeader(parentBlockHeader)
+            .number(parentBlockHeader.getNumber() + 1)
+            .timestamp(timestampForNextBlock)
+            .parentHash(parentBlockHeader.getHash())
+            .blockHeaderFunctions(new MainnetBlockHeaderFunctions())
+            .buildBlockHeader();
+    return getByBlockHeader(nextBlockHeader);
   }
+
+  Optional<BigInteger> getChainId();
+
+  String listMilestones();
+
+  void putMilestone(
+      final boolean isTimestampMilestone,
```

Comment:
really don't like this boolean, maybe enum would make it clearer

Reply:
Used inheritance instead as discussed https://github.com/hyperledger/besu/pull/5310/commits/816137a4a75f8430beda3d0e1df35093b6fdadfa